### PR TITLE
Emit 'pre' and 'routed' events

### PR DESCRIPTION
**Pull request by yunong**
WIP

### CODE REVIEW

**Comment by yunong** [comment]
@rajatkumar @DonutEspresso please take a look.

File: docs/pages/components/server.md [hunk 1]
```diff
@@ -294,6 +294,46 @@ Note that when the server automatically responds with a
 NotFound/MethodNotAllowed/VersionNotAllowed, this event will still be fired.
 
 
+## pre
+
+Before each request has been routed, a `pre` event is fired. This event can be
+hooked into handle audit logs and other metrics. Since this event fires
+*before* routing has occured, it will firing regardless of whether the route is
```
Comment (rajatkumar) [inline docs/pages/components/server.md]: `it will fire`

File: docs/pages/components/server.md [hunk 2]
```diff
@@ -294,6 +294,46 @@ Note that when the server automatically responds with a
...
+Note that when the server automatically responds with a
+NotFound/MethodNotAllowed/VersionNotAllowed, this event will still be fired.
+
+
+## routed
+
+A `routed` event is fired after a request has been routed by the router, but
+before handlers specific to that route has run.
+
+The sginature for the `routed` event is as follows:
```
Comment (rajatkumar) [inline docs/pages/components/server.md]: `signature`

File: test/server.test.js [hunk 3]
```diff
@@ -2742,3 +2742,116 @@ test('should show debug information', function (t) {
...
+        res.send();
+        next();
+    });
+
+    var count = 0;
+    SERVER.once('pre', function (req, res) {
+        t.ok(req);
+        t.ok(res);
+
+        if (++count === 2) {
```
Comment (DonutEspresso) [inline test/server.test.js]: Please remove counts, event emitters are synchronous and there is no race condition here against the pre or route events.

**Pull request merged**